qiskit/qpy/objects/circuits.py

```diff
@@ -0,0 +1,625 @@
+    qubit_indices = dict(enumerate(circuit.qubits))
+    clbit_indices = dict(enumerate(circuit.clbits))
+
+    # Load Arguments
+    for _qarg in range(num_qargs):
+        qarg_raw = file_obj.read(INSTRUCTION_ARG_SIZE)
+        qarg = struct.unpack(INSTRUCTION_ARG_PACK, qarg_raw)
+        if qarg[0].decode("utf8") == "c":
+            raise TypeError("Invalid input carg prior to all qargs")
+        qargs.append(qubit_indices[qarg[1]])
+    for _carg in range(num_cargs):
+        carg_raw = file_obj.read(INSTRUCTION_ARG_SIZE)
+        carg = struct.unpack(INSTRUCTION_ARG_PACK, carg_raw)
+        if carg[0].decode("utf8") == "q":
+            raise TypeError("Invalid input qarg after all qargs")
+        cargs.append(clbit_indices[carg[1]])
+
+    # Load Parameters
+    for _ in range(num_params):
+        type_key, data = read_binary(file_obj)
+        if TypeKey.is_number(type_key) or TypeKey.is_variable(type_key):
+            value = loads_parameter_value(type_key, data)
+        elif type_key == TypeKey.STRING:
+            value = data.decode("utf8")
+        else:
+            raise TypeError("Invalid parameter type: %s" % type_key)
+        params.append(value)
+
+    # Load Gate object
+    gate_class = None
+    if gate_name in ("Gate", "Instruction"):
+        inst_obj = _parse_custom_instruction(custom_instructions, gate_name, params)
+        inst_obj.condition = condition_tuple
+        if label_size > 0:
+            inst_obj.label = label
+        circuit._append(inst_obj, qargs, cargs)
+        return
+    elif gate_name in custom_instructions:
+        inst_obj = _parse_custom_instruction(custom_instructions, gate_name, params)
+        inst_obj.condition = condition_tuple
+        if label_size > 0:
+            inst_obj.label = label
+        circuit._append(inst_obj, qargs, cargs)
+        return
+    elif hasattr(library, gate_name):
+        gate_class = getattr(library, gate_name)
+    elif hasattr(circuit_mod, gate_name):
+        gate_class = getattr(circuit_mod, gate_name)
+    elif hasattr(extensions, gate_name):
+        gate_class = getattr(extensions, gate_name)
+    elif hasattr(quantum_initializer, gate_name):
+        gate_class = getattr(quantum_initializer, gate_name)
+    else:
+        raise AttributeError("Invalid instruction type: %s" % gate_name)
+    if gate_name == "Initialize":
+        gate = gate_class(params)
+    else:
+        if gate_name == "Barrier":
+            params = [len(qargs)]
+        gate = gate_class(*params)
+    gate.condition = condition_tuple
+    if label_size > 0:
+        gate.label = label
+    if not isinstance(gate, Instruction):
+        circuit.append(gate, qargs, cargs)
+    else:
+        circuit._append(gate, qargs, cargs)
+
+
+def _parse_custom_instruction(custom_instructions, gate_name, params):
+    (type_str, num_qubits, num_clbits, definition) = custom_instructions[gate_name]
+    if type_str == "i":
```

Comment:
`TypeKey` enum strategy doesn't work here because `i` is also used for Integer (because key is not scoped). This logic imperfection looks bit awkward (but removing enum makes pulse qpy code really hard to read because there are many kinds of operand classes).